### ISSUE TITLE
SQlite Error 11 'database disk image is malformed'

### ISSUE THREAD

**Issue by pmouldsramp** [dotnet/efcore#33630]
### Version Info

EFCore / Microsoft.Data.Sqlite Version: 3.1.17
.Net Version: Xamarin.Android (MonoAndroid v10)

We have been using EFCore with SQlite in a Xamarin Android app, and we've started running into issues with this malformed database error randomly when reading or writing. 

It started off as only a handful of cases, and we thought it was caused by the app being killed during a device restart, however it has been happening an increasing amount. We have ~200 devices running this app and in the last month ~25 devices have run into this error. 
Once the issue occurs it appears permanent (i.e. the database is actually malformed), and the only way to fix it is to delete and restore the db. Sometimes it only occurs once per device, and sometimes it happens consistently on the same device.

We are running the database in WAL mode, but not sure what difference this makes to the safety.
```
var connectionString = !_connectionString.StartsWith("Filename=") ? "Filename=" + _connectionString : _connectionString;
var conn = new SqliteConnection(connectionString);
conn.Open();

// Enable write-ahead log
var cmd = conn.CreateCommand();
cmd.CommandText = "PRAGMA journal_mode = 'wal'";
cmd.ExecuteNonQuery();

return conn;
```

### Stacktrace
```
Exception Found:
Type: Microsoft.EntityFrameworkCore.DbUpdateException
Message: An error occurred while updating the entries. See the inner exception for details.
Source: mscorlib
Stacktrace:   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync (Microsoft.EntityFrameworkCore.Storage.IRelationalConnection connection, System.Threading.CancellationToken cancellationToken) [0x00219] in <198e7898c63e4d45b5c9f1fb14cc1493>:0 
  at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync (System.Collections.Generic.IEnumerable`1[T] commandBatches, Microsoft.EntityFrameworkCore.Storage.IRelationalConnection connection, System.Threading.CancellationToken cancellationToken) [0x00222] in <198e7898c63e4d45b5c9f1fb14cc1493>:0 
  at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync (System.Collections.Generic.IEnumerable`1[T] commandBatches, Microsoft.EntityFrameworkCore.Storage.IRelationalConnection connection, System.Threading.CancellationToken cancellationToken) [0x00390] in <198e7898c63e4d45b5c9f1fb14cc1493>:0 
  at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync (System.Collections.Generic.IList`1[T] entriesToSave, System.Threading.CancellationToken cancellationToken) [0x00088] in <7129d9bae44b47969ca04dc4a34d5e52>:0 
  at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync (Microsoft.EntityFrameworkCore.DbContext _, System.Boolean acceptAllChangesOnSuccess, System.Threading.CancellationToken cancellationToken) [0x0010a] in <7129d9bae44b47969ca04dc4a34d5e52>:0 
  at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync (System.Boolean acceptAllChangesOnSuccess, System.Threading.CancellationToken cancellationToken) [0x000e9] in <7129d9bae44b47969ca04dc4a34d5e52>:0 
  at Cleanaway.Core.Engine.Repo.RepoBase`1[TEntity].SaveChangesAsync (Microsoft.EntityFrameworkCore.DbContext context, System.String errorDescription) [0x00125] in <af472cd547a34166b239661b5bd97278>:0 
  Inner Exception Found:
  Type: Microsoft.Data.Sqlite.SqliteException
  Message: SQLite Error 11: 'database disk image is malformed'.
  Source: mscorlib
  Stacktrace:   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC (System.Int32 rc, SQLitePCL.sqlite3 db) [0x0006d] in <472c4d99df014496b15861dcf509cece>:0 
  at Microsoft.Data.Sqlite.SqliteDataReader.NextResult () [0x0017b] in <472c4d99df014496b15861dcf509cece>:0 
  at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader (System.Data.CommandBehavior behavior) [0x000d4] in <472c4d99df014496b15861dcf509cece>:0 
  at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReaderAsync (System.Data.CommandBehavior behavior, System.Threading.CancellationToken cancellationToken) [0x00007] in <472c4d99df014496b15861dcf509cece>:0 
  at Microsoft.Data.Sqlite.SqliteCommand.ExecuteDbDataReaderAsync (System.Data.CommandBehavior behavior, System.Threading.CancellationToken cancellationToken) [0x00011] in <472c4d99df014496b15861dcf509cece>:0 
  at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync (Microsoft.EntityFrameworkCore.Storage.RelationalCommandParameterObject parameterObject, System.Threading.CancellationToken cancellationToken) [0x00278] in <198e7898c63e4d45b5c9f1fb14cc1493>:0 
  at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync (Microsoft.EntityFrameworkCore.Storage.RelationalCommandParameterObject parameterObject, System.Threading.CancellationToken cancellationToken) [0x004c9] in <198e7898c63e4d45b5c9f1fb14cc1493>:0 
  at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync (Microsoft.EntityFrameworkCore.Storage.RelationalCommandParameterObject parameterObject, System.Threading.CancellationToken cancellationToken) [0x0057b] in <198e7898c63e4d45b5c9f1fb14cc1493>:0 
  at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync (Microsoft.EntityFrameworkCore.Storage.IRelationalConnection connection, System.Threading.CancellationToken cancellationToken) [0x000cc] in <198e7898c63e4d45b5c9f1fb14cc1493>:0 
```

We also sometimes get Sqlite error 10 Disk I/O and Sqlite error 26 Not a Database.

We've tried to reproduce the issue on a test device by constantly reading/writing to the database but haven't had any success.

**Comment by ericsink** [comment]
Nothing specifically familiar here.  Looks like a corrupted database file of course.  I can't help but notice that the version of Microsoft.Data.Sqlite is very old, and it's not clear what version of SQLite itself is involved.  Android devices at one point made it easy to have "the multiple SQLite problem", which was a straightforward way to cause a corruption.  But these are just guesses.

**Comment by roji** [comment]
Good catch @ericsink; @pmouldsramp note that 3.1 is out of support.

**Comment by cincuranet** [comment]
@ajcvickers @roji is there something left to do? Without at least semi-reliable repro, it's hard to guess where the problem comes from.

@pmouldsramp did you try newer version? Did it happen (ever) again?

**Comment by ajcvickers** [comment]
@cincuranet We need a repro to do anything here.

**Comment by pmouldsramp** [comment]
@cincuranet I think I can close this issue now. We weren't able to update to a newer version, but the issue was not with efcore.

FYI, we found the issue was caused by a hardware vendor library we are also using. Apparently concurrent access to a serial port device from this library could somehow affect the database. We would see lots of avc: denied errors from Android re the database file, and the serial port and then the database would become corrupted. Preventing the concurrent access stopped the issue.
